### ISSUE TITLE
Cannot pair SF30 Pro

### ISSUE THREAD

**Issue by AndehX** [pixelcircuits/GBConsole#4]
Just finished building this, but I can't for the life of me get the SF30 Pro to pair.  The documentation doesn't really give enough info on how to properly pair it.  There are numerous different pairing modes for this controller, and I have no idea which one im supposed to use.  I've tried all 4 and none of them will pair with the GBConsole.

**Comment by pixelcircuits** [comment]
Only the SN30 and SN30 Pro are currently supported. The SF30 Pro should still work though if paired outside of the app or simply plugged in via USB. I think it's looking for dinput mode (which I believe is the default).

**Comment by AndehX** [comment]
ah man, I don't have an OTG adapter.  I'll have to order one.  I was hoping I'd just be able to pair the controller since I figured the SF30 was basically the same as the SN30.  Do you plan on adding support for other common bluetooth controllers like PS3/PS4 controller?

**Comment by pixelcircuits** [comment]
Try pairing the controller manually (maybe using bluetoothctrl) if you're comfortable with command line. I would assume the other 8bitdo controllers would be easy to add support for. I only added support for the cheaper controllers as a cost saving measure. PS3/PS4 controller support would be the next obvious addition due to how common they are.

**Comment by AndehX** [comment]
I'm really not familiar with the command line stuff, and im mainly building this for my girlfriend for her birthday as an convenient way to play her Game Boy games on the TV.  If you could at least add PS3/PS4 controller support, that would be alot more useful than the 8bitdo controllers, as she (and probably most people) already have these controllers lying around.

**Comment by AndehX** [comment]
Ok, I got hold of an SN30 controller, but im having problems getting it to work properly.  It has 4 pairing modes.  I can only get it paired with the "Start+X" and "Start+B" modes, both of which have issues with the Dpad in the menu.  "Start+X" dpad does not work at all in the menu, and "Start+B" only right on the Dpad works.  Also the keyboard also only seems to register right too, and there's no audio when running a game.

Edit:  Ok, using a different SD card seemed to give better results.  Now my keyboard works, and I have sound, but I cannot get the controller to pair up at all now...

**Comment by pixelcircuits** [comment]
The Raspberry Pi Zero wants a class 10 SD card, but I'm not sure how that would be effecting bluetooth. I just double checked with my SN30 and you'll need to set it to dinput mode (Start+B) and switch it into pairing mode by holding down select until the light starts blinking fast. Make sure to keep the controller next to the console and give it plenty of time to pair (it took up to 40s sometimes when I tried it). Hope this helps. I'm leaving this issue open until I update the documentation to be this detailed.

**Comment by AndehX** [comment]
Thanks for your reply.  I tried a different (class 10) SD card and it seems to have fixed the issue.  I've also got the SN30 working now too.  So everything seems to be working fine now.

Is there any way to add artwork for games that are not in the master packs?  I'd like to add artwork for Pokemon Crystal Clear.

**Comment by pixelcircuits** [comment]
I'm glad you've got everything working! 
In theory, you can add an entry for any game not recognized to the master list (data/GameBoyColor.json) and then add any relevant artwork to the libretro art repository (/home/pi/libretro/gbc) using the same name you provided on the list. The "crc" and "sha1" properties are not currently used and can be left empty or set to something random for now. The hard part will be getting the start hash ("sha1_1k") and size of the rom file ("size"). You can use the command line tool I created ([GBTool](https://github.com/pixelcircuits/GBTool)) to get these values with the "info" command.

If you find any games that are not recognized and are not rom hacks, please let me know.

**Comment by AndehX** [comment]
quick question, how do I run GBTool?  I've got it downloaded and installed, i'm just not sure how to run it from the terminal
I tried typing "GBTool info" but I keep getting the message "-bash: GBTool: command not found"

**Comment by AndehX** [comment]
that sorted it!  For some reason I can't run GBA games though.  I copied the gba_bios.bin file to the root of the boot partition, but the GBA intro screen just repeats constantly when I try to load a game

**Comment by pixelcircuits** [comment]
Every GBA game does the same thing? or is it a specific game?

**Comment by AndehX** [comment]
well I've only tried one so far, Pokemon Emerald

**Comment by pixelcircuits** [comment]
Pokemon Emerald works for me. Did you install everything from source? or did you use the pre built image?

**Comment by AndehX** [comment]
I used the prebuilt image.  I'm not familiar enough with all this to do it all manually

**Comment by pixelcircuits** [comment]
I'm not sure why the emulator isn't working. Maybe the bios file is bad? Hopefully in the future I can remove the need for a bios file

**Comment by AndehX** [comment]
well I tried 2 different bios files, but I guess I could look around and try as many as I can find, and see if any of them work.

I just need to copy them to the root of the boot partition right?  and that will overwrite the previous one I used?

**Comment by AndehX** [comment]
Ok, i've tried multiple different games and different gba_bios.bin files but the result always seems to be the same.

I've also just spent the last 3 hours building it from scratch instead of using the prebuilt image, and I got the exact same result.  All GBA games I try seem to just boot loop.

I should also point out that GBA emulation does not work at all without the gba_bios file.  All I get is a solid white screen when attempting to load a game.

EDIT:  Ok I've worked out what's going on.  It's not ripping the rom from the cart properly.  I tried copying the roms off the SD card and running them in VisualBoyAdvance and they don't work.  So there's an issue with the GBConsole ripping roms from GBA carts.  (Ripping GB/GBC games works fine)
It seems whatever GBA game I rip, the filesize is always 32,768Kb.

**Comment by AndehX** [comment]
I don't mean to be a bother, but do you know what's going on?

**Comment by pixelcircuits** [comment]
The ROM is being recognized correctly though? Correct album art and title is being displayed? I think there might be a few issues with reproduction carts if that's what you're using. The [GBTool](https://github.com/pixelcircuits/GBTool) can be used to do more analysis (use the 'info' and 'debug' commands and check if the reported data is correct). I think it's safe to say it might be a hardware issue so I think that's where most effort should be placed. 

Have you tried getting a correct ROM dump somewhere to test if at least the save data is being ripped correctly?

**Comment by AndehX** [comment]
Yeah the ROM is recognised correctly, it shows the art normally.  The cart I was using was an InsideGadgets RTC GBA cart.  I ripped the ROM and save data from the cart using a GBxCart RW, and then manually copied them to the GBConsole, and it loads the game fine like that.

**Comment by pixelcircuits** [comment]
So currently, not many reproduction carts are supported. I think the InsideGadgets gameboy carts may work, but I don't think any of the GBA carts have been tested. Do you have any genuine GBA carts to test?

**Comment by AndehX** [comment]
I don't have any genuine GBA carts, no.  I've tried InsideGadgets GBA carts and chinese bootleg carts, and both produce the same result.  I mostly only need it to play various rom hacks, im not really interested in playing original carts.
